homeassistant/helpers/config_validation.py

```diff
@@ -518,18 +535,90 @@ def ensure_list_csv(value: Any) -> Sequence:
     return ensure_list(value)
 
 
-def deprecated(key):
-    """Log key as deprecated."""
+def deprecated(key: str,
+               replacement_key: Optional[str] = None,
+               invalidation_version: Optional[str] = None,
+               default: Optional[Any] = None):
+    """
+    Log key as deprecated and provide a replacement (if exists).
+
+    Expected behavior:
+        - Outputs the appropriate deprecation warning if key is detected
+        - Processes schema moving the value from key to replacement_key
+        - Processes schema changing nothing if only replacement_key provided
+        - No warning if only replacement_key provided
+        - No warning if neither key nor replacement_key are provided
+            - Adds replacement_key with default value in this case
+        - Once the invalidation_version is crossed, raises vol.Invalid if key
+        is detected
+    """
     module_name = inspect.getmodule(inspect.stack()[1][0]).__name__
 
-    def validator(config):
+    if replacement_key and invalidation_version:
+        warning = ("The '%(key)s' option (with value '%(value)s') is"
+                   " deprecated, please replace it with '%(replacement_key)s'."
+                   " This option will become invalid in version"
+                   " %(invalidation_version)s.")
+    elif replacement_key:
+        warning = ("The '%(key)s' option (with value '%(value)s') is"
+                   " deprecated, please replace it with '%(replacement_key)s'."
+                   )
+    elif invalidation_version:
+        warning = ("The '%(key)s' option (with value '%(value)s') is"
+                   " deprecated, please remove it from your configuration."
+                   " This option will become invalid in version"
+                   " %(invalidation_version)s.")
+    else:
+        warning = ("The '%(key)s' option (with value '%(value)s') is"
+                   " deprecated, please remove it from your configuration.")
+
+    def check_for_invalid_version(value: Optional[Any]):
+        """Raise error if current version has reached invalidation."""
+        if not invalidation_version:
+            return
+
+        major_version, minor_version, patch_version = \
+            map(int, invalidation_version.split('.', 2))
+        # PATCH_VERSION can be a string when running dev/betas, int otherwise
+        running_patch_version = (int(PATCH_VERSION.split('.', 1)[0])
+                                 if isinstance(PATCH_VERSION, str)
+                                 else PATCH_VERSION)
+        if (MAJOR_VERSION >= major_version
```

Comment:
I think we can use `parse_version` from `pkg_resources` to parse and compare versions.

https://stackoverflow.com/a/21065570

Reply:
That makes things much simpler, thanks!